### DISCUSSION TITLE
Alt+Space not working in Windows

### DISCUSSION THREAD

**Discussion by Agarvaen** [AvaloniaUI/Avalonia#14518] [Q&A] [answered]
Can't open a menu with Alt+Space in Windows. 
Works fine in Ubuntu.
Is their a workaround?

**Comment by gehongyan** [comment, marked as answer]
### Reason

It seems that Alt+Space was broken by #6592.

https://github.com/AvaloniaUI/Avalonia/blob/b3a8ba25cf3e6f43fe89b8df91eb339bc361e482/src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs#L158-L162

If I changed the code to:

```csharp
const char spaceChar = ' ';
if ((SysCommands)wParam == SysCommands.SC_KEYMENU && HighWord(ToInt32(lParam)) <= 0 && ToInt32(lParam) != spaceChar)
    return IntPtr.Zero;
```

Alt+Space shall work now, but I am not a pro at interacting with Win32 API, so I am not sure whether this change will result in potential issues.

![image](https://github.com/AvaloniaUI/Avalonia/assets/21241496/deb0eebe-4f74-4e91-be11-b9c62fc963e6)

---

### Workaround

If you accept to use Alt+Space `KeyUpEvent` to manually show the window menu (the originally working behavior seems like `KeyDownEvent`), here is a workaround:

- Allow unsafe blocks in your csproj:
  ```xml
  <Project Sdk="Microsoft.NET.Sdk">
      <PropertyGroup>
          <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
      </PropertyGroup>
  </Project>
  ```

- Import Win32 API:
  ```csharp
  using System;
  using System.Runtime.InteropServices;

  internal static partial class UnmanagedMethods
  {
      [LibraryImport("user32.dll", EntryPoint = "DefWindowProcW")]
      public static partial IntPtr DefWindowProc(IntPtr hWnd, uint msg, IntPtr wParam, IntPtr lParam);
  }
  ```

- Defines a method that attaches the handler to `KeyUpEvent` in `MainWindow`:
  ```csharp
  using System;
  using Avalonia.Controls;
  using Avalonia.Input;
  using Avalonia.Interactivity;

  private void TryRegisterWindowsMenu()
  {
      if (!OperatingSystem.IsWindows()) return;
      if (GetTopLevel(this)?.TryGetPlatformHandle() is not { } handle) return;
      const uint wmSysCommandMessage = 0x0112;
      const nint sysCommandsScKeyMenu = 0xF100;
      const nint spaceChar = ' ';
      AddHandler(KeyUpEvent, (sender, args) =>
      {
          if (args.Key != Key.Space || args.KeyModifiers == KeyModifiers.None) return;
          UnmanagedMethods.DefWindowProc(handle.Handle, wmSysCommandMessage, sysCommandsScKeyMenu, spaceChar);
      }, RoutingStrategies.Tunnel);
  }
  ```

- Invoke `TryRegisterWindowsMenu()` in the constructor of `MainWindow`.

Now, the window menu shall pop up when space with any key modifiers is pressed and then released.

---

### References

- #6592 
- #6597 
- [WM_SYSCOMMAND message (Winuser.h) - Win32 apps | Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/menurc/wm-syscommand#remarks)

**Comment by Agarvaen** [comment]
Thanks.  Strange decision to disable system handling of Alt/F10. Looks like a quick fix, without fixing the root of the problem.
